tfhe/src/core_crypto/commons/ciphertext_modulus.rs

```diff
@@ -49,52 +64,101 @@ impl<'de, Scalar: UnsignedInteger> serde::Deserialize<'de> for CiphertextModulus
             )));
         }
 
-        Ok(CiphertextModulus(thing.modulus, PhantomData))
+        let res = if thing.modulus == 0 {
+            CiphertextModulus {
+                inner: CiphertextModulusInner::Native,
+                _scalar: PhantomData,
+            }
+        } else {
+            CiphertextModulus {
+                inner: CiphertextModulusInner::Custom(NonZeroU128::new(thing.modulus).ok_or(
+                    serde::de::Error::custom(
+                        "Got zero modulus for CiphertextModulusInner::Custom variant",
+                    ),
+                )?),
+                _scalar: PhantomData,
+            }
+        };
+        Ok(res.canonicalize())
     }
 }
 
 impl<Scalar: UnsignedInteger> CiphertextModulus<Scalar> {
     pub const fn new_native() -> Self {
-        Self(0, PhantomData)
+        Self {
+            inner: CiphertextModulusInner::Native,
+            _scalar: PhantomData,
+        }
     }
 
     pub const fn try_new_power_of_2(exponent: usize) -> Result<Self, &'static str> {
         if exponent > Scalar::BITS {
             Err("Modulus is bigger than the maximum value of the associated Scalar type")
         } else {
-            let modulus = match 1u128.checked_shl(exponent as u32) {
-                Some(modulus) => modulus,
-                None => 0,
+            let res = match 1u128.checked_shl(exponent as u32) {
+                Some(modulus) => {
+                    let non_zero_modulus = match NonZeroU128::new(modulus) {
+                        Some(val) => val,
+                        None => {
+                            return Err(
+                                "Got zero modulus for CiphertextModulusInner::Custom variant",
+                            )
+                        }
+                    };
```

Comment:
If I'm correct, here, If we got `Some(modulus)` that means modulus is 100% guaranteed not te be zero,
So we could use the unchecked_new (not in favor of thats) or `.expect(msg)` the `NonZeroU128::new` result
instead of matching on it

but can we panic in const fn ? 🤔

Reply:
I don't think we can panic and that's why I wrote it this way

Reply:
I'll recheck

Reply:
looks like here the const panic is fine if we don't format a string, I would expect the compiler to strip it anyway as we are not supposed to get there, I just went conservative this time as my previous CiphertextModulus struct was a bit error prone (and likely has some UB as M1 tests did not run properly)